lib/spack/spack/test/stage.py

```diff
@@ -699,3 +699,81 @@ def test_stage_constructor_with_path(self, tmpdir):
         testpath = str(tmpdir)
         with Stage('file:///does-not-exist', path=testpath) as stage:
             assert stage.path == testpath
+
+    def test_diystage_path_none(self):
+        """Ensure DIYStage for path=None behaves as expected."""
+        stage = DIYStage(None)
+        assert stage.path is None
+        assert stage.source_path is None
+
+        # Order doesn't really matter for DIYStage
+        stage.create()  # Only sets the flag value
+        assert stage.created
+
+        stage.cache_local()  # Only outputs a message
+        stage.fetch()  # Only outputs a message
+        stage.check()  # Only outputs a message
+        stage.expand_archive()  # Only outputs a message
+
+        with pytest.raises(TypeError):
+            assert stage.expanded  # os.path.exists() fails on None
+
+        with pytest.raises(SystemExit):
+            stage.restage()  # Calls tty.die()
+
+        stage.destroy()  # A no-op
+
+        assert stage.path is None
+        assert stage.source_path is None
+
+    def test_diystage_path_invalid(self):
+        """Ensure DIYStage for an invalid path behaves as expected."""
+        path = '/path/does/not/exist'
+        stage = DIYStage(path)
```

Comment:
I would also omit the rest of this test and just ensure that `DIYStage(path)` raises some reasonable error (probably a subclass of `StageError`) if the `path` isn't a directory.

If you do this, you can simplify your `expanded` implementation, too -- just add `expanded = True` as a property in `DIYStage`.